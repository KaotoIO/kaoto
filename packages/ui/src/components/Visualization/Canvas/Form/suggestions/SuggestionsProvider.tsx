import { useSuggestionRegistry } from '@kaoto/forms';
import { FunctionComponent, PropsWithChildren, useEffect } from 'react';
import { propertiesSuggestionProvider } from './suggestions/properties.suggestions';
import { simpleLanguageSuggestionProvider } from './suggestions/simple-language.suggestions';
import { sqlSyntaxSuggestionProvider } from './suggestions/sql.suggestions';

export const SuggestionRegistrar: FunctionComponent<PropsWithChildren> = ({ children }) => {
  const suggestionRegistryContext = useSuggestionRegistry();

  useEffect(() => {
    suggestionRegistryContext?.registerProvider(propertiesSuggestionProvider);
    suggestionRegistryContext?.registerProvider(simpleLanguageSuggestionProvider);
    suggestionRegistryContext?.registerProvider(sqlSyntaxSuggestionProvider);

    return () => {
      suggestionRegistryContext?.unregisterProvider(propertiesSuggestionProvider.id);
      suggestionRegistryContext?.unregisterProvider(simpleLanguageSuggestionProvider.id);
      suggestionRegistryContext?.unregisterProvider(sqlSyntaxSuggestionProvider.id);
    };
  }, [suggestionRegistryContext]);

  return <>{children}</>;
};
