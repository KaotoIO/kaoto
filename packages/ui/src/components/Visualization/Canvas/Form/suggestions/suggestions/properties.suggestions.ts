import { SuggestionProvider } from '@kaoto/forms';

export const propertiesSuggestionProvider: SuggestionProvider = {
  id: 'properties-suggestion-provider',
  appliesTo: (_propName, schema) => schema.type === 'string',
  getSuggestions: async (word, _context) => {
    word ??= 'foo';

    return [
      // https://camel.apache.org/manual/using-propertyplaceholder.html
      {
        value: `RAW(${word})`,
        description: `Use the RAW value of the '${word}' property`,
      },
      {
        value: `{{${word}}}`,
        description: `Use '${word}' as a property reference`,
      },
      {
        value: `{{${word}:default}}`,
        description: `Use '${word}' with a default value as a property reference`,
      },
      {
        value: `{{?${word}}}`,
        description: `Use '${word}' as an optional property reference`,
        group: 'Property references',
      },
      {
        value: `{{!${word}}}`,
        description: `Use '${word}' as a negated property reference`,
        group: 'Property references',
      },

      // https://camel.apache.org/manual/using-propertyplaceholder.html#_using_property_placeholder_functions
      {
        value: `{{env:${word}}}`,
        description: `Use the '${word}' OS environment variable`,
        group: 'Property functions',
      },
      {
        value: `{{sys:${word}}}`,
        description: `Use the '${word}' Java JVM system property`,
        group: 'Property functions',
      },
      {
        value: `{{service:${word.toUpperCase()}}}`,
        description: `Use the ${word.toUpperCase()} service, defined using the service naming idiom. Services are expected to be defined as OS ${word.toUpperCase()}_SERVICE_HOST and ${word.toUpperCase()}_SERVICE_PORT environment variables.`,
        group: 'Property functions',
      },
      {
        value: `{{service.name:${word.toUpperCase()}}}`,
        description: `Use the ${word.toUpperCase()} service hostname`,
        group: 'Property functions',
      },
      {
        value: `{{service.port:${word.toUpperCase()}}}`,
        description: `Use the ${word.toUpperCase()} service port`,
        group: 'Property functions',
      },

      // https://camel.apache.org/manual/using-propertyplaceholder.html#_using_kubernetes_property_placeholder_functions
      {
        value: `{{configmap:${word}}}`,
        description: `Use the ${word} ConfigMap. It should be in the format 'configmap:configmap-name/configmap-key'`,
        group: 'Kubernetes',
      },
      {
        value: `{{configmap:${word}}}:defaultValue`,
        description: `Use the ${word} ConfigMap with a default value`,
        group: 'Kubernetes',
      },
      {
        value: `{{configmap-binary:${word}}}`,
        description: `Use the binary ${word} ConfigMap`,
        group: 'Kubernetes',
      },
      {
        value: `{{secret:${word}}}`,
        description: `Use the ${word} Secret. It should be in the format 'secret:secret-name/secret-key'`,
        group: 'Kubernetes',
      },
      {
        value: `{{secret-binary:${word}}}`,
        description: `Use the binary ${word} Secret.`,
        group: 'Kubernetes',
      },
    ];
  },
};
