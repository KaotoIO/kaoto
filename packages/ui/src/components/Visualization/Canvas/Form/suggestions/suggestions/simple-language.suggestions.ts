import { Suggestion, SuggestionProvider } from '@kaoto/forms';
import { CamelCatalogService, CatalogKind } from '../../../../../../models';
import { IMetadataApi } from '../../../../../../providers';

const SIMPLE_LANGUAGE_ACTIVATED_FIELDS = [
  '#.message', // This is a special case for the Log EIP
  'simple.expression',
];

// This provider suggests syntax from: https://camel.apache.org/components/next/languages/simple-language.html
export const getSimpleLanguageSuggestionProvider = (metadata?: IMetadataApi['getSuggestions']): SuggestionProvider => {
  return {
    id: 'simple-language-suggestion-provider',
    appliesTo: (propName, schema) => SIMPLE_LANGUAGE_ACTIVATED_FIELDS.includes(propName) && schema.type === 'string',
    getSuggestions: async (word, context) => {
      const normalizedWord = word !== '' ? word : 'name';
      const normalizedWordLowercase = normalizedWord.toLowerCase();

      const simpleLangFunctions = CamelCatalogService.getComponent(CatalogKind.Function, 'simple') ?? {};
      const {
        body,
        ['variable.name']: variableName,
        ['header.name']: headerName,
        ['env.name']: envName,
        ...remainingFunctions
      } = simpleLangFunctions;

      const suggestions: Suggestion[] = [
        { value: `\${variable.${normalizedWord}}`, description: variableName?.description },
        { value: `\${header.${normalizedWord}}`, description: headerName?.description },
      ];

      const bodySuggestion: Suggestion = { value: '${body}', description: body?.description };
      if (word === '') {
        suggestions.unshift(bodySuggestion);
      } else {
        suggestions.push({ ...bodySuggestion, group: 'Simple Language' });
      }

      const environmentVariables = (await metadata?.('env', word, context)) ?? [];
      const environmentVariablesSuggestions = environmentVariables.map((item) => ({
        value: `\${env:${item.value}}`,
        description: `Use the '${item.value}' OS environment variable`,
        group: 'Simple Language: Environment variables',
      }));
      if (environmentVariablesSuggestions.length > 0) {
        suggestions.push(...environmentVariablesSuggestions);
      } else {
        suggestions.push({ value: `\${env.${normalizedWord}}`, description: envName?.description });
      }

      return suggestions.concat(
        Object.entries(remainingFunctions)
          .filter(([name]) => word === '' || name.toLowerCase().startsWith(normalizedWordLowercase))
          .map(([name, func]) => ({
            value: `\${${name}}`,
            description: func.description,
            group: 'Simple Language',
          })),
      );
    },
  };
};
