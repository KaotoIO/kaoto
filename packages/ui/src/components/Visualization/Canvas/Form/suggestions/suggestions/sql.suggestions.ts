// Example of SQL syntax suggestion provider

import { SuggestionProvider } from '@kaoto/forms';
import { simpleLanguageSuggestionProvider } from './simple-language.suggestions';

// This provider suggests syntax from: https://camel.apache.org/components/next/languages/sql.html
export const sqlSyntaxSuggestionProvider: SuggestionProvider = {
  ...simpleLanguageSuggestionProvider,
  id: 'sql-syntax-suggestion-provider',
  appliesTo: (propName, schema) => propName === '#.parameters.query' && schema.type === 'string',
  getSuggestions: async (word, _context) => {
    const suggestionWord = word !== '' ? word : 'foo';

    return [
      {
        value: `:#\${variable.${suggestionWord}}`,
        description: `Reference the '${suggestionWord}' variable`,
      },
      {
        value: `:#\${header.${suggestionWord}}`,
        description: `Reference the '${suggestionWord}' header`,
      },
    ];
  },
};
