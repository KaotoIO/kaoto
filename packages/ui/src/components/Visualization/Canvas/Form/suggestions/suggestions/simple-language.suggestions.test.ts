import { simpleLanguageSuggestionProvider } from './simple-language.suggestions';

describe('Properties Suggestions', () => {
  it.each([
    { propertyName: 'test', schema: { type: 'string' }, expected: true },
    { propertyName: 'test', schema: { type: 'number' }, expected: false },
    { propertyName: 'test', schema: { type: 'object' }, expected: false },
    { propertyName: 'test', schema: { type: 'array' }, expected: false },
    { propertyName: 'test', schema: { type: 'boolean' }, expected: false },
  ] as const)('should apply to string properties', ({ propertyName, schema, expected }) => {
    const applies = simpleLanguageSuggestionProvider.appliesTo(propertyName, schema);
    expect(applies).toBe(expected);
  });

  it('should return suggestions', async () => {
    const word = 'test';
    const suggestions = await simpleLanguageSuggestionProvider.getSuggestions(word, {
      inputValue: 'test example',
      cursorPosition: 0,
      propertyName: 'test-property',
    });

    expect(suggestions).toMatchSnapshot();
  });
});
