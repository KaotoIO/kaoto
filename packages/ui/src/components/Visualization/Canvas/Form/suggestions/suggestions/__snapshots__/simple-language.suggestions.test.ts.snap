// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properties Suggestions should return suggestions 1`] = `
[
  {
    "description": "Use the body of the message",
    "group": "Simple Language",
    "value": "\${body}",
  },
  {
    "description": "Use the 'test' header of the message",
    "group": "Simple Language",
    "value": "\${header.test}",
  },
  {
    "description": "Use the 'test' variable of the message",
    "group": "Simple Language",
    "value": "\${variable.test}",
  },
]
`;
